File: src/systems/save.js
// Save / load to localStorage, with a versioned envelope.

const KEY = 'ironworks_save_v1';

export function saveGame(game) {
  try {
    const data = {
      v: 1, when: Date.now(),
      money: game.money, time: game.time.t, day: game.time.day,
      player: game.player.serialize(),
      factory: game.factory.serialize(),
      floors: [...game.floors.keys()],
      research: game.research.serialize(),
      missions: game.missions.serialize(),
      employees: game.employees.serialize(),
      market: { price: game.market.price },
      contracts: { active: game.contracts.active, offers: game.contracts.offers, completed: game.contracts.completed },
      plot: game.world.plot, expansions: game.expansions,
      nodes: [...game.world.nodes.entries()].map(([k, n]) => [k, n.amount]),
      stats: game.stats,
    };
    localStorage.setItem(KEY, JSON.stringify(data));
    return true;
  } catch (err) {
    console.error('save failed', err);
    return false;
  }
}

export function hasSave() {
  try { return !!localStorage.getItem(KEY); } catch { return false; }
}

export function loadGame(game) {
  try {
    const raw = localStorage.getItem(KEY);
    if (!raw) return false;
    const d = JSON.parse(raw);
    game.money = d.money;
    game.time.t = d.time; game.time.day = d.day;
    game.player.deserialize(d.player);
    game.research.deserialize(d.research);
    game.floors = new Map((d.floors || []).map((k) => [k, true]));
    game.factory.deserialize(d.factory);
    game.missions.deserialize(d.missions);
    game.employees.deserialize(d.employees);
    if (d.market) Object.assign(game.market.price, d.market.price);
    if (d.contracts) {
      game.contracts.active = d.contracts.active || [];
      game.contracts.offers = d.contracts.offers || [];
      game.contracts.completed = d.contracts.completed || 0;
    }
    if (d.plot) Object.assign(game.world.plot, d.plot);
    game.expansions = d.expansions || 0;
    for (const [k, amount] of d.nodes || []) {
      const n = game.world.nodes.get(k);
      if (n) n.amount = amount; else if (amount <= 0) game.world.nodes.delete(k);
    }
    if (d.stats) Object.assign(game.stats, d.stats);
    game.factory.dirtyPower = true;
    return true;
  } catch (err) {
    console.error('load failed', err);
    return false;
  }
}

export function clearSave() { try { localStorage.removeItem(KEY); } catch {} }

File: ironworks/systems/save.py
"""Save / load to a JSON file next to the game."""
import json
import os

SAVE_PATH = os.path.join(os.path.expanduser('~'), '.ironworks_save.json')
VERSION = 1


def save_game(game, path=SAVE_PATH):
    try:
        data = dict(
            v=VERSION,
            money=game.money, time=game.time['t'], day=game.time['day'],
            expansions=game.expansions,
            player=game.player.serialize(),
            factory=game.factory.serialize(),
            floors=['%d,%d' % k for k in game.floors],
            research=game.research.serialize(),
            missions=game.missions.serialize(),
            employees=game.employees.serialize(),
            market=dict(price=game.market.price),
            contracts=dict(active=game.contracts.active, offers=game.contracts.offers,
                           completed=game.contracts.completed),
            plot=game.world.plot,
            nodes=[['%d,%d' % k, n.amount] for k, n in game.world.nodes.items()],
            stats=game.stats,
        )
        with open(path, 'w') as fh:
            json.dump(data, fh)
        return True
    except Exception as err:                                   # pragma: no cover
        print('save failed:', err)
        return False


def has_save(path=SAVE_PATH):
    return os.path.exists(path)


def load_game(game, path=SAVE_PATH):
    try:
        with open(path) as fh:
            d = json.load(fh)
        game.money = d['money']
        game.time['t'] = d['time']
        game.time['day'] = d['day']
        game.expansions = d.get('expansions', 0)
        game.player.deserialize(d['player'])
        game.research.deserialize(d['research'])
        game.floors = {tuple(int(v) for v in k.split(',')): True for k in d.get('floors', [])}
        game.factory.deserialize(d['factory'])
        game.missions.deserialize(d['missions'])
        game.employees.deserialize(d['employees'])
        if d.get('market'):
            game.market.price.update(d['market']['price'])
        if d.get('contracts'):
            game.contracts.active = d['contracts'].get('active', [])
            game.contracts.offers = d['contracts'].get('offers', [])
            game.contracts.completed = d['contracts'].get('completed', 0)
        if d.get('plot'):
            game.world.plot.update(d['plot'])
        for key, amount in d.get('nodes', []):
            k = tuple(int(v) for v in key.split(','))
            n = game.world.nodes.get(k)
            if n:
                n.amount = amount
            elif amount <= 0:
                game.world.nodes.pop(k, None)
        if d.get('stats'):
            game.stats.update(d['stats'])
        game.factory.dirty_power = True
        return True
    except Exception as err:                                   # pragma: no cover
        print('load failed:', err)
        return False


def clear_save(path=SAVE_PATH):
    try:
        os.remove(path)
    except OSError:
        pass
